---
title: "Strain transmission among cohabitants"
author: "Jing Yang"
date: "2024-09-07"
output: html_document
---

## Generating matrix containing multiple sequence alignment results from StrainPhlan

Turn fasta format file into matrix, each element of the matrix is a base, rows are sites, columns are samples.

```{r}
setwd("F:/AAAA/Others/WH_healthy_OP/Summary/WHOP_Manuscript/strain_reanalysis")
file <- list.files("F:/AAAA/Others/WH_healthy_OP/data/strain/strainphlan/full_msas_69_species/",pattern = "*.fa")
dat <- lapply(file,function(x) read.table(paste("F:/AAAA/Others/WH_healthy_OP/data/strain/strainphlan/full_msas_69_species/",x,sep = ""),sep = "\t",header = F))
names(dat)=file
species=read.table("F:/AAAA/Others/WH_healthy_OP/data/strain/strainphlan/profiled_successed_species_69",sep = "\t",header = F)
species=species$V1

fasta_into_matrix <- function(data){
  data=data$V1
  data_names=data[seq(1,length(data),2)]
  data_names=gsub(">","",data_names)
  
  data_seq=data[seq(2,length(data),2)]
  samples=unique(data_names)
  data_seq_matrix=matrix(0,nrow = length(data_seq),ncol = nchar(data_seq[1]))
  for(i in 1:nrow(data_seq_matrix)){
    for(j in 1:ncol(data_seq_matrix)){
      data_seq_matrix[i,j]=substring(data_seq[i],j,j)
    }
  }
  rownames(data_seq_matrix)=data_names
  data_seq_matrix=as.data.frame(t(data_seq_matrix))
  return(data_seq_matrix)
}

fasta_matrix=list()

for(k in 1:length(species)){
  print(species[k])
  data=lapply(dat[grep(species[k],names(dat))], fasta_into_matrix)
  names(data)=grep(species[k],file,value = T)
  
  total_length=lapply(data, nrow) %>% unlist() %>% sum()
  total_sample=lapply(data, colnames) %>% unlist() %>% unique()
  data1=matrix("0",nrow = total_length,ncol = length(total_sample)) %>% as.data.frame()
  names(data1)=total_sample
  
  for(j in total_sample){
    seq=c()
    for(i in grep(species[k],file,value = T)){
      samples=names(data[[i]])
      seq_len=nrow(data[[i]])
      if(length(which(samples==j))!=0){
        seq=c(seq,data[[i]][,j] %>% as.character())
      }else{seq=c(seq,rep("-",seq_len))}
      
    }
    data1[,j]=seq
  }
  
  fasta_matrix[[k]]=data1
}

names(fasta_matrix)=species

strain_mincov3_full <- fasta_matrix

rm(data,dat,file,fasta_into_matrix,fasta_matrix,data1,
   i,j,k,samples,seq,seq_len,species,tmp,total_length,total_sample)
save(strain_mincov3_full,file="full_fasta_matrix.RData",version = 2)
```

## Calculating SNP distance among all the samples

```{r}
load("full_fasta_matrix.RData")
load("F:/AAAA/Others/WH_healthy_OP/metadata_1046.RData")

strain_mincov3_full=strain_mincov3_full[which(lapply(strain_mincov3_full, nrow)!=0)]


flattenCorrMatrix <- function(cormat) {
  ut <- upper.tri(cormat)
  data.frame( sample1 = rownames(cormat)[row(cormat)[ut]],
              sample2 = rownames(cormat)[col(cormat)[ut]], dist =(cormat)[ut])
}


compute_dist <- function(fasta_matrix){
  print("Begin")
  tmp1=ape::as.DNAbin(as.matrix(t(fasta_matrix)))
  tmp1=ape::dist.dna(tmp1,model = "raw",pairwise.deletion = T,
                     as.matrix = T) %>% flattenCorrMatrix() %>%as.data.frame()
  tmp1=tmp1[is.nan(tmp1$dist)==F,]
  tmp1=tmp1[order(tmp1$dist),]
  
  ###total_compare_length###time-consuming##
  tmp1$total_compare_length=apply(tmp1,1,function(y){
    length(intersect(which(fasta_matrix[,y[1]]!="-"),which(fasta_matrix[,y[2]]!="-")))})
  

  f1=tmp1$sample1
  f2=tmp1$sample2
  f1=data.frame(sample=f1,family=meta_1[f1,"family"])
  f2=data.frame(sample=f2,family=meta_1[f2,"family"])
  tmp1[,"type"]=ifelse(f1$family==f2$family,"cohabitation","non-cohabitation")

  print("sample_pairs:")
  print(dim(tmp1))
  print(tail(tmp1))
  print(str(tmp1))
  return(tmp1)
  
}

strain_dist <- lapply(strain_mincov3_full, compute_dist)
rm(strain_mincov3_full)
save(strain_dist,file="strain_distance_all_pairs.RData")

```

## Strain distance among cohabitation pairs and non-cohabitation pairs

overlap base length above 1000 per pair

```{r}
##intra-pairs: cohabiting pairs
intra_pairs <- function(total_pairs){
  intra_pair=total_pairs[total_pairs$type=="cohabitation",]# & total_pairs$total_compare_len>=1000
  print("end")
  return(intra_pair)}

strain_intra_pair <- lapply(strain_dist,intra_pairs)

for(i in 1:length(strain_intra_pair_rm10)){
  if(nrow(strain_intra_pair_rm10[[i]])!=0){
    strain_intra_pair_rm10[[i]][,"species"]=names(strain_intra_pair_rm10)[i]
  }
  
}

strain_intra_pair=Reduce(rbind,strain_intra_pair)


strain_intra_pair$pairs=apply(strain_intra_pair,1,function(x){paste(x[6],paste(min(x[1],x[2]),max(x[1],x[2]),sep = "_"),sep = ":")})



####inter-pairs: non-cohabiting pairs
inter_pairs <- function(total_pairs){
  inter_pair=total_pairs[total_pairs$type=="non-cohabitation" &
                         total_pairs$total_compare_len>=1000,]
  print("end")
  return(inter_pair)}
strain_inter_pair=lapply(strain_dist,inter_pairs)

for(i in 1:length(strain_inter_pair)){
  if(nrow(strain_inter_pair[[i]])!=0){
    strain_inter_pair[[i]][,"species"]=names(strain_inter_pair)[i]
  }
}
strain_inter_pair=Reduce(rbind,strain_inter_pair)

strain_intra_pair$pairs=apply(strain_intra_pair,1,function(x){paste(min(x[1],x[2]),max(x[1],x[2]),sep = "_")})
tt=strain_intra_pair
strain_inter_pair_sub=strain_inter_pair[0,]
for(i in unique(tt$species) %>% as.character()){
  intra_samples=strsplit(tt$pairs[tt$species==i],split = "_") %>% unlist() %>% unique()
  kk=strain_inter_pair[strain_inter_pair$species==i,]
  intra_inter=kk[(kk$sample1  %in% intra_samples) & (kk$sample2 %in%  intra_samples),]
  strain_inter_pair_sub=rbind(strain_inter_pair_sub,intra_inter)
}
strain_inter_pair_sub$pairs=apply(strain_inter_pair_sub,1,function(x){paste(min(x[1],x[2]),max(x[1],x[2]),sep = "_")})


```

## Data visualization of comparing genotype distance between different inter-person relationships

```{r}
####Figure S3A
species_table=table(strain_intra_pair$species) %>% as.data.frame() %>% .[order(.[,2],decreasing = T),]
names(species_table)[2]="intra_numbers"

strain_39_heatmap=data.frame(matrix(NA,39,141)) ##39 species & 141 cohabitating pairs

rownames(strain_39_heatmap)=species_table$Var1
colnames(strain_39_heatmap)=unique(strain_intra_pair$pairs)

fam_relation=read.csv("F:/AAAA/Others/WH_healthy_OP/all_family_relation.csv")
rownames(fam_relation)=fam_relation$pairs
fam_141=data.frame(pairs=unique(strain_intra_pair$pairs))
fam_141$type=fam_relation[fam_141$pairs,"type"]
fam_141=fam_141[order(fam_141$type),]
fam_141$type=factor(fam_141$type,levels = c("others","offspring2","offspring1","siblings","spouse"),
                    labels = c("Others","Grandparent_child","Parent_child","Sibling","Spouse"))

strain_39_heatmap=strain_39_heatmap[,fam_141$pairs]

for(i in 1:nrow(strain_39_heatmap)){
  for(j in 1:ncol(strain_39_heatmap)){
   tmp=strain_intra_pair$dist[strain_intra_pair$species==rownames(strain_39_heatmap)[i] & strain_intra_pair$pairs==colnames(strain_39_heatmap)[j]]
    strain_39_heatmap[i,j]=ifelse(length(tmp)!=0,tmp,NA) 
  }
}

###inter distance stat##
total_inter_summ=as.data.frame(table(strain_inter_pair_sub$species))
rownames(total_inter_summ)=total_inter_summ$Var1
total_inter_summ=total_inter_summ[rownames(strain_39_heatmap),]

strain_39_heatmap_inter=dcast(strain_inter_pair_sub[,c("species","pairs","dist")],species~pairs,value.var = "dist")
rownames(strain_39_heatmap_inter)=strain_39_heatmap_inter$species
strain_39_heatmap_inter=strain_39_heatmap_inter[,-1]
strain_39_heatmap_inter=strain_39_heatmap_inter[rownames(strain_39_heatmap),]
rownames(strain_39_heatmap_inter)=rownames(strain_39_heatmap)

load("F:/AAAA/Others/WH_healthy_OP/kraken_species_rela.RData")
kraken_39_species_rela=kraken_species_rela[rownames(strain_39_heatmap),meta_1$sampleID]

annot2=rowAnnotation(median_in_inter=apply(strain_39_heatmap_inter,1,function(x){median(x %>% na.omit)}),
                     aver_abun=apply(kraken_39_species_rela,1,mean) %>% as.numeric(),
                     prevalance=apply(kraken_39_species_rela,1,function(x){length(which(x>0.001))}),
                     intra_pair_numbers=species_table[rownames(strain_39_heatmap),"intra_numbers"],
                     col=list(aver_abun=colorRamp2(c(0,0.1), c("#FFE8C8","#FFA501")),
                              prevalance=colorRamp2(c(0,1028),c("#C7D3E9","#4F81BD")),
                              intra_pair_numbers=colorRamp2(c(10,100), c("white","#6C41FD")),
                              median_in_inter=colorRamp2(c(0.08,0.06,0.04,0.035,0.03,0.025,0.02,0.01,0.005,0.0025,0),
                                            c("#002C60","#2662AA","#3F93C0","#8BC3DC","#CEE4F0","#F7F7F7","#FAD9C3","#F1A482","#D65F4D","#B21828","#64001F"))),
                     show_legend=T,show_annotation_name=F)


annot3=columnAnnotation(relation_types=fam_141$type,
                        col=list(relation_types=c(Others="#2082E1",Grandparent_child="#F1CD33",Parent_child="#9D9D9D",Spouse="#D77570",Sibling="#A59133")),
                        show_legend=T,show_annotation_name=F)
  
Heatmap(strain_39_heatmap %>% as.matrix(),
           col = colorRamp2(c(0.08,0.06,0.04,0.035,0.03,0.025,0.02,0.01,0.005,0.0025,0),
                            c("#002C60","#2662AA","#3F93C0","#8BC3DC","#CEE4F0","#F7F7F7","#FAD9C3","#F1A482","#D65F4D","#B21828","#64001F")),
           name = "Hamming distance",
           show_heatmap_legend = T,
           rect_gp = gpar(col = "grey90"),
           column_title = "141 cohabitating pairs",
           column_title_side = "bottom",
           row_labels = gsub("_"," ",rownames(strain_39_heatmap)),
           row_names_side = "left",
           cluster_rows=F,cluster_columns = F,
           show_row_names = T,show_column_names = F,
           right_annotation =annot2,
           top_annotation = annot3
)


##Figure3A
tt=strain_intra_pair
tt=rbind(tt,strain_inter_pair_sub)

ggplot(tt,aes(x=dist))+ 
  geom_histogram(aes(y = stat(width*density),fill=type),alpha=0.5,color="grey30",position = "identity")+
  scale_y_continuous(labels = scales::percent)+
  scale_fill_manual(values = c("#985E12","#0C564E"))+
    labs(x="Hamming distance",y="Percentage")+theme_test()+
  theme(legend.position = "top")

ggplot(tt)+
  stat_ecdf(aes(x=dist,color=type),geom = "step",size=1)+
  #geom_density(aes(x=dist,fill=type),alpha=0.6)+
  scale_color_manual(values = c("#985E12","#0C564E"))+
  labs(x="Hamming distance",y="Cumulative fraction")+
  theme_test()+
  theme(legend.position = "none")

###K-S test
ks.test(tt$dist[tt$type=="cohabitation"],tt$dist[tt$type=="non-cohabitation"],alternative = "two.sided")
length(which(tt$dist[tt$type=="cohabitation"]<0.025))/length(which(tt$type=="cohabitation"))
length(which(tt$dist[tt$type=="non-cohabitation"]<0.025))/length(which(tt$type=="non-cohabitation"))


##Figure3B
tmp=table(strain_inter_pair_sub$species) %>% as.data.frame()
names(tmp)[2]="inter_numbers"
species_table=left_join(species_table,tmp)
rm(tmp)
species_table=species_table[order(species_table$intra_numbers,decreasing = T),]
species_table=species_table[!is.na(species_table$inter_numbers),]

######intra & inter每个物种的boxplot比较
tt=tt[tt$species %in% (species_table$Var1 %>% as.character()),]
tt$species=factor(tt$species,levels = rev(rownames(strain_39_heatmap)))
ggplot(data = tt,aes(y = dist,x = species,fill=type))+
  geom_boxplot(alpha=0.8,outlier.size = 0.1,outlier.color = "grey50")+
  stat_compare_means(label = "p.signif",hide.ns = T, method = "wilcox.test")+
  scale_fill_manual(values = c("#0C564E","#985E12"))+
  labs(x="",y="Hamming distance")+theme_test()+
  theme(axis.text.x = element_text(angle = 0),legend.position = "top")+
  ggtitle("")+
  coord_flip()


######Figure3G distance among different familt relations####
tt=left_join(tt,fam_141,by="pairs")
tt$type.y=as.character(tt$type.y)
tt$type.y[is.na(tt$type.y)]="non-cohabitation"
tt=tt[tt$type.y!="Others",]

tt$type.y=factor(tt$type.y,levels = c("non-cohabitation","Grandparent_child","Parent_child","Spouse","Sibling"),
                    labels = c("Non-cohabitation","Grandparent_child","Parent_child","Spouse","Sibling"))

ggplot(data = tt,aes(y = dist,x = type.y,fill=type.y))+
  geom_boxplot(alpha=0.8,outlier.size = 0.1,outlier.color = "grey50")+
  stat_compare_means(comparisons = list(c("Grandparent_child","Sibling"),
                                        c("Grandparent_child","Spouse"),
                                        c("Grandparent_child","Parent_child"),
                                        c("Non-cohabitation","Sibling"),
                                        c("Spouse","Non-cohabitation"),
                                        c("Non-cohabitation","Parent_child")
                                        ))+
  scale_fill_manual(values = c("#85AAA6","#EFC000FF","#868686FF","#CD534CFF","#8F7700FF"))+
  labs(x="",y="Hamming distance")+theme_test()+
  theme(axis.text.x = element_text(angle = 0),legend.position = "right")


###Figure S5F strate for different species
ggplot(data = tt[tt$species %in% c("Neisseria_subflava","Neisseria_flavescens","Prevotella_melaninogenica","Prevotella_histicola", "Prevotella_jejuni","Veillonella_dispar","Haemophilus_parainfluenzae"),],aes(y = dist,x = type.y,fill=type.y))+
  geom_boxplot(alpha=0.8,outlier.size = 0.1,outlier.color = "grey50")+
  stat_compare_means(comparisons = list(c("Spouse","Non-cohabitation"),
                                        c("Sibling","Non-cohabitation")
                                        ))+
  scale_fill_manual(values = c("#85AAA6","#EFC000FF","#868686FF","#CD534CFF","#8F7700FF"))+
  labs(x="",y="Hamming distance")+theme_test()+
  theme(axis.text.x = element_text(angle = 0),legend.position = "right")+
  facet_wrap(.~species,nrow = 2,scales = "free")
```




## Defining strain transmission events for each species using StrainPhlan

```{r}
######cohabitation/non-cohabitation
distance_sub=seq(0,max(strain_intra_pair$dist[strain_intra_pair$species %in% as.character(species_table$Var1)]),0.001)[-1]

strain_distance_heatmap=data.frame(matrix(0,nrow = length(species_table$Var1),ncol = length(distance_sub)))
rownames(strain_distance_heatmap)=species_table$Var1
colnames(strain_distance_heatmap)=distance_sub
strain_distance_heatmap_p=strain_distance_heatmap

for (i in rownames(strain_distance_heatmap)) {
  for(j in 1:length(distance_sub)){
    a=length(which(strain_intra_pair$dist[strain_intra_pair$species==i]<=distance_sub[j]))
    b=length(which(strain_inter_pair_sub$dist[strain_inter_pair_sub$species==i]<=distance_sub[j]))
    c=species_table$intra_numbers[species_table$Var1==i]
    d=species_table$inter_numbers[species_table$Var1==i]
    if(a==0){
      strain_distance_heatmap[i,j]=NA
    }else if(a!=0 & b==0){
      strain_distance_heatmap[i,j]=(a/c)/((b+1)/(d+1))
    }else{strain_distance_heatmap[i,j]=(a/c)/(b/d)}
    
    test=fisher.test(matrix(c(a,b,(c-a),(d-b)),nrow = 2))
    strain_distance_heatmap_p[i,j]=test$p.value
    
  }
}

########################define transmission events####
####the maximum genetic distance ensured that the probability of sharing the same strain in cohabiting pairs was two times higher than that in non-cohabiting pairs, with significance in Fisher's exact test####
tmp=apply(strain_distance_heatmap,1,function(x){return(which(x>2))})
tmp1=apply(strain_distance_heatmap_p,1,function(x){return(which(x<0.05))})

or_p=c()
for(k in 1:34){
  or_p[k]=intersect(tmp[[k]] %>% as.numeric(),tmp1[[k]] %>% as.numeric()) %>% max()
}
or_p_thre=data.frame(species=names(tmp),threshold=distance_sub[or_p])

#####Number transmission events####
or_p_thre$trans=NA
for(i in 1:34){
  or_p_thre[i,"trans"]=length(which(strain_intra_pair$dist[strain_intra_pair$species==or_p_thre$species[i]]<=or_p_thre$threshold[i]))
}
all(or_p_thre$species==species_table$Var1)
or_p_thre$total_num=species_table$intra_numbers
or_p_thre$trans_rate=or_p_thre$trans/or_p_thre$total_num
or_p_thre$trans_rate[or_p_thre$trans_rate==0]=NA
or_p_thre$trans[or_p_thre$trans==0]=NA

####FigureS5C
annot1=rowAnnotation(intra_pair_numbers=species_table$intra_numbers,
                     threshold=or_p_thre$threshold,
                     trans_num=or_p_thre$trans,
                     trans_rate=or_p_thre$trans_rate,
                     col=list(intra_pair_numbers=colorRamp2(c(2,101),c("#F9F5FF","#6C41FD")),
                              threshold=colorRamp2(c(0.006,0.038),c("#1C9E77","white")),
                              trans_rate=colorRamp2(c(0.04,1),c("white","#D95F02")),
                              trans_num=colorRamp2(c(1,23),c("white","#FFE200"))),
                     show_legend=T,show_annotation_name=F)


p1=Heatmap(log2(strain_distance_heatmap) %>% as.matrix(),
           col = colorRamp2(c(-0.5,-0.4,-0.3,-0.2,-0.1,0,2,4,6,8,10),
                            c("#002C60","#2662AA","#3F93C0","#8BC3DC","#CEE4F0","#F7F7F7","#FAD9C3","#F1A482","#D65F4D","#B21828","#64001F")),
           name = "log2(cohabitation%/non-cohabitation%)",
           color_space = c(-0.5,-0.4,-0.3,-0.2,-0.1,0,2,4,6,8,10),
           show_heatmap_legend = T,
           rect_gp = gpar(col = "grey90"),
           column_title = "Distance range",
           column_title_side = "bottom",
           column_labels = distance_sub,
           column_names_rot = 90,
           column_names_gp = gpar(fontsize = 8),
           row_labels = gsub("_"," ",rownames(strain_distance_heatmap)),
           row_names_side = "left",
           cluster_rows=F,cluster_columns = F,
           show_row_names = T,show_column_names = T,
           right_annotation =annot1,
           cell_fun = function(j,i,x,y,w, h, fill){
             if(strain_distance_heatmap_p[i,j]<0.05){
               grid.text("*",x,y)}
             if((is.na(strain_distance_heatmap[i,j])==F) & (strain_distance_heatmap[i,j]>2)){
               grid.text(".",x,y)
             }
             }
           )
p1

```

## Statistical analysis after defining transmission events

```{r}
###Figure S5E compare species abundance among transmitted pairs and non-tranmitted pairs###
strain_intra_pair1=merge(strain_intra_pair,or_p_thre)
strain_intra_pair1$trans=ifelse(strain_intra_pair1$dist<=strain_intra_pair1$threshold,
                                     "trans","non-trans")
strain_intra_pair1=strain_intra_pair1[!is.na(strain_intra_pair1$threshold),]
trans_samples=unique(c(strain_intra_pair1$sample1[strain_intra_pair1$trans=="trans"],strain_intra_pair1$sample2[strain_intra_pair1$trans=="trans"]))
nontrans_samples=unique(c(strain_intra_pair1$sample1[strain_intra_pair1$trans=="non-trans"],strain_intra_pair1$sample2[strain_intra_pair1$trans=="non-trans"]))
nontrans_samples=setdiff(nontrans_samples,trans_samples)

abundance_trans=kraken_species_rela[unique(strain_intra_pair1$species),c(trans_samples,nontrans_samples)]

plotdata=strain_intra_pair1[,c("sample1","sample2","species","trans")]
plotdata$max_abun=apply(plotdata,1,function(x){return(max(abundance_trans[x[3],x[1]],abundance_trans[x[3],x[2]]))})
plotdata$avg_abun=apply(plotdata,1,function(x){return(mean(c(abundance_trans[x[3],x[1]],abundance_trans[x[3],x[2]])))})

ggplot(plotdata,aes(x=species,y=avg_abun,color=trans))+
  geom_boxplot(width=0.5,outlier.size = 0.1)+
  scale_color_lancet()+
  stat_compare_means(label = "p.signif",hide.ns = F, method = "wilcox.test")+
  theme_test()+xlab("")+ylab("Average relative abundance per pair")+coord_flip()

plotdata1=aggregate(avg_abun~species+trans,plotdata,mean)
plotdata2=aggregate(avg_abun~species+trans,plotdata,sd)
names(plotdata1)[3]="mean"
plotdata1$sd=plotdata2$avg_abun
rm(plotdata2)
plotdata1[is.na(plotdata1)]=0

plotdata1$trans=factor(plotdata1$trans,levels = c("non-trans","trans"),labels = c("non-transmission","transmission"))
plotdata1$species=factor(plotdata1$species,levels = rev(c("Neisseria_subflava",
                                                      "Neisseria_flavescens",
                                                      "Prevotella_melaninogenica",
                                                      "Prevotella_histicola",
                                                      "Prevotella_sp_oral_taxon_306",
                                                      "Prevotella_pallens",
                                                      "Prevotella_jejuni",
                                                      "Veillonella_atypica",
                                                      "Veillonella_dispar",
                                                      "Megasphaera_micronuciformis",
                                                      "Prevotella_salivae",
                                                      "Veillonella_tobetsuensis",
                                                      "Neisseria_perflava",
                                                      "Prevotella_nigrescens",
                                                      "Fusobacterium_periodonticum")))
ggplot(plotdata1,aes(x=species,y=mean,color=trans))+
  geom_point(position=position_dodge(0.8),shape=1,size=2)+
  geom_errorbar(aes(ymin=mean-sd, ymax=mean+sd), width=.6,position=position_dodge(0.8),size=1)+
  scale_color_lancet()+
  theme_test()+xlab("")+ylab("Average relative abundance per pair")+coord_flip()
```

## Figure3D strain transmission rate compare with oral(Nature)
```{r}
nature_data=read_excel("F:/AAAA/Others/WH_healthy_OP/papers/菌株传递/person-person transmission landscape/41586_2022_5620_MOESM3_ESM.xlsx",sheet="Table S30")
nature_oral_taxa=read_excel("F:/AAAA/Others/WH_healthy_OP/papers/菌株传递/person-person transmission landscape/41586_2022_5620_MOESM3_ESM.xlsx",sheet="Table S5")

nature_oral_taxa$species=apply(nature_oral_taxa %>% as.data.frame,1,function(x){return(strsplit(x[3] %>% as.character(),split="\\|")%>% unlist() %>% grep("s__",.,value = T))})
nature_oral_taxa$species=gsub("s__","",nature_oral_taxa$species)

thre=or_p_thre[is.na(or_p_thre$threshold)==F,]
setdiff(thre$species,nature_oral_taxa$species)
nature_oral_taxa$species[nature_oral_taxa$SGB=="SGB9450_group"]="Neisseria_subflava"

thre1=left_join(thre,nature_oral_taxa[,c("species","SGB")])
thre1=left_join(thre1,nature_data[,c("SGB","SGB_household_transmissibility")])
thre1=thre1[is.na(thre1$SGB)==F & thre1$total_num>=5,]
thre1$SGB_household_transmissibility=as.numeric(thre1$SGB_household_transmissibility)

ggplot(thre1,aes(x=trans_rate,y=SGB_household_transmissibility))+geom_point()+#aes(color=species)
  geom_text(aes(label=species))+
  geom_smooth(method = "lm",color="navyblue",fill=NA)+
  stat_cor(method = "spearman",label.x = 0.1,label.y = 0.7)+theme_test()+#scale_color_simpsons()+
  xlab("Transmssion rate in OP")+ylab("Transmission rate in oral")
```

## Co-transmitted species
```{r}
###Figure3E lower: Correlation of species genotype distance between different species in the same cohabitating pairs
species_distance_cor=data.frame(matrix(NA,15,15))
rownames(species_distance_cor)=or_p_thre$species[!is.na(or_p_thre$threshold)]
colnames(species_distance_cor)=or_p_thre$species[!is.na(or_p_thre$threshold)]
species_distance_p=species_distance_cor

for(i in 1:15){
  for(j in 1:15){
    if(i!=j){
       tmp=dcast(strain_intra_pair1[strain_intra_pair1$species %in% c(rownames(species_distance_cor)[i],colnames(species_distance_cor)[j]),c("species","pairs","dist")],species~pairs,value.var = "dist")
       k1=which(is.na(tmp[1,-1])==F)
       k2=which(is.na(tmp[2,-1])==F)
       if(length(intersect(k1,k2))>3){
         test=cor.test(tmp[1,-1] %>% as.numeric(),tmp[2,-1] %>% as.numeric(),method = "spearman")
         species_distance_cor[i,j]=test$estimate
         species_distance_p[i,j]=test$p.value
       }

    }
    
    }
    
}
species_distance_cor[is.na(species_distance_cor)]=0
species_distance_p[is.na(species_distance_p)]=1
#species_distance_p=ifelse(species_distance_p<0.05,"*","")
#species_distance_p[is.na(species_distance_p)]=""

corrplot(corr=species_distance_cor %>% as.matrix(),
         na.label = " ",
         type = "lower",
         tl.col = "black",
         is.corr = T,
         p.mat = species_distance_p %>% as.matrix(),
         sig.level = 0.05,
         col.lim = c(-1,1),
         insig = "label_sig",
         col = colorRampPalette(c("#67001F", "#B2182B", "#D6604D", "#F4A582","#FDDBC7", "#FFFFFF","#CEE4F0","#8BC3DC","#3F93C0","#2662AA","#002C60"))(100) %>%rev(),
         addCoef.col = NULL,
         tl.srt = 90,
         method = "square"#,
         #number.digits = 2,
         #number.font = 1
         )

###Figure3E upper: co-transmission events identified by hypergeometric test
species_distance_cor=data.frame(matrix(NA,15,15))
rownames(species_distance_cor)=or_p_thre$species[!is.na(or_p_thre$threshold)]
colnames(species_distance_cor)=or_p_thre$species[!is.na(or_p_thre$threshold)]
species_distance_p=species_distance_cor

for(i in 1:15){
  for(j in 1:15){
    if(i!=j){
      k1=strain_intra_pair1$pairs[strain_intra_pair1$species==rownames(species_distance_cor)[i] & 
                                         strain_intra_pair1$trans=="trans"]
      k2=strain_intra_pair1$pairs[strain_intra_pair1$species==colnames(species_distance_cor)[j] & 
                                         strain_intra_pair1$trans=="trans"]
      n=length(intersect(k1,k2))
      species_distance_cor[i,j]=n
      test=fisher.test(matrix(c(n,(length(k1)-n),(length(k2)-n),(141-length(union(k1,k2)))),nrow=2))
      species_distance_p[i,j]=test$p.value
    }
    
  }
}

corrplot(corr=species_distance_cor %>% as.matrix(),
         na.label = " ",
         type = "upper",
         tl.col = "black",
         is.corr = F,
         p.mat = species_distance_p %>% as.matrix(),
         insig = "label_sig",
         col = colorRampPalette(c("#67001F", "#B2182B", "#D6604D", "#F4A582",
                                  "#FDDBC7", "#FFFFFF"))(100) %>%rev(),
         #col.lim = c(0,1),
         addCoef.col = "black",
         tl.srt = 90,
         method = "square",
         number.digits = 2,
         number.font = 1)


## Figure3F
flattenCorrMatrix = function(cormat, pmat) {
ut = upper.tri(cormat) 
data.frame( row = rownames(cormat)[row(cormat)[ut]], column = rownames(cormat)[col(cormat)[ut]], cor =(cormat)[ut], p = pmat[ut] )
}
tmp=flattenCorrMatrix(cormat = species_distance_cor,pmat = species_distance_p)
tmp$g1=apply(tmp,1,function(x){strsplit(x[1],split="_") %>% unlist() %>% .[1]})
tmp$g2=apply(tmp,1,function(x){strsplit(x[2],split="_") %>% unlist() %>% .[1]})
tmp$is.same.genus=(tmp$g1==tmp$g2)
tmp$simul_trans=ifelse(tmp$p<0.05,"yes","no")

table(tmp[,c("is.same.genus","simul_trans")]) %>% fisher.test()

tmp$simul_trans=factor(tmp$simul_trans,levels = c("yes","no"))
tmp %>%
  ggplot() +
  geom_mosaic(aes(x = product(is.same.genus), fill = simul_trans)) +
  scale_y_continuous(breaks = seq(0, 1.4, by = 0.2)) +
  ylim(0,1.05)+
  xlab("is.same.genus") +
  ylab("Proportion") +
  theme_test()+scale_fill_manual(values = c("#DF8F44FF","#374E55FF"))


```

## Figure3H transmission events in different family relationships
```{r}
plotdata=left_join(strain_intra_pair1,fam_141,by="pairs")
plotdata$type.y=as.character(plotdata$type.y)
plotdata=plotdata[plotdata$type.y!="Others",]
table(plotdata[,c("trans","type.y")])
##fisher test p-value
fisher.test(matrix(c(51,138,12,55),nrow = 2))
fisher.test(matrix(c(41,67,12,55),nrow = 2))
fisher.test(matrix(c(4,8,12,55),nrow = 2))

fisher.test(matrix(c(41,67,51,138),nrow = 2))
fisher.test(matrix(c(4,8,51,138),nrow = 2))

fisher.test(matrix(c(4,8,41,67),nrow = 2))

plotdata$trans=factor(plotdata$trans,levels = c("non-trans","trans"),labels = c("non-transmission","transmission"))
plotdata$type.y=factor(plotdata$type.y,levels = c("Grandparent_child","Parent_child","Spouse","Sibling"))

plotdata %>%
  ggplot() +
  geom_mosaic(aes(x = product(type.y), fill = trans),alpha=0.7) +
  scale_y_continuous(breaks = seq(0, 1.4, by = 0.2)) +
  ylim(0,1.05)+
  xlab("Relation types") +
  ylab("Proportion") +
  theme_test()+
  scale_fill_lancet()
```

